Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+        {
+            return Directory.Exists(directory)
+                ? Directory.EnumerateFiles(directory, $"{LogFilePrefix}*{LogFileSuffix}", SearchOption.AllDirectories).ToArray()
+                : CollectionUtilities.EmptyArray<string>();
+        }
+
+        /// <summary>
+        /// Returns all paths recorded in the <paramref name="filePath"/> file.
+        /// </summary>
+        /// <remarks>
+        /// Those paths are expected to be absolute paths of files/directories that were written to by the previous build.
+        ///
+        /// NOTE: this method does not validate the recorded paths in any way.  That means that each returned string may be
+        ///   - a path pointing to an absent file
+        ///   - a path pointing to a file
+        ///   - a path pointing to a directory.
+        /// 
+        /// NOTE: unless the log file was corrupted (i.e., it was produced by an instance of this class):
+        ///   - the strings in the returned enumerable are all legal paths
+        ///   - the returned collection does not contain any duplicates
+        /// </remarks>
+        public static IEnumerable<string> ReadRecordedPathsFromSharedOpaqueOutputLog(string filePath)
+        {
+            using (var fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete))
+            using (var bxlReader = new BuildXLReader(stream: fileStream, debug: false, leaveOpen: false))
+            {
+                string nextString = null;
+                while ((nextString = ReadStringOrNull(bxlReader)) != null)
+                {
+                    yield return nextString;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Same as <see cref="ReadRecordedPathsFromSharedOpaqueOutputLog(string)"/> except that all exceptions are wrapped in <see cref="BuildXLException"/>
+        /// </summary>
+        public static string[] ReadRecordedPathsFromSharedOpaqueOutputLogWrapExceptions(string filePath)
+        {
+            try
+            {
+                return ReadRecordedPathsFromSharedOpaqueOutputLog(filePath).ToArray();
+            }
+            catch (Exception e)
+            {
+                throw new BuildXLException("Failed to read from shared opaque output log", e);
+            }
+        }
+
+        /// <summary>
+        /// <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>
+        /// </summary>
+        public bool RecordFileWrite(PathTable pathTable, string absolutePath)
+            => RecordFileWrite(pathTable, AbsolutePath.Create(pathTable, absolutePath));
+
+        /// <summary>
+        /// Records that the file at location <paramref name="path"/> was written to.
+        /// 
+        /// Returns whether the path was recorded or skipped.
+        /// </summary>
+        /// <returns>
+        /// <code>true</code> if <paramref name="path"/> was recorded, <code>false</code> 

```

Comment:
This is confusing, could you add a remark saying it should be recorded if the path is within any given root directory?

Reply:
```suggestion
        /// <code>true</code> if <paramref name="path"/> is within any given root directory (<see cref="RootDirectories"/>) and hence was recorded; <code>false</code> otherwise.
```